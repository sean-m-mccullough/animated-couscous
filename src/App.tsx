import Button from '@mui/material/Button';
import Card from '@mui/material/Card';
import CardActions from '@mui/material/CardActions';
import CardContent from '@mui/material/CardContent';
import CardHeader from '@mui/material/CardHeader';
import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import Contributions from './components/Contributions/Contributions';
import ContributionEdit from './components/Contribution/Edit/Edit';
import { Status } from './types/contribution';
import { getSelected } from './selectors/contributions';
import { edit } from './App.actions';

const App: React.FC = () => {
  const selected = useSelector(getSelected);
  const actionableSelection = !!selected && selected.status === Status.Pending;

  const dispatch = useDispatch();

  function handleEdit(): void {
    dispatch(edit());
  }

  return (
    <>
      <Card elevation={0}>
        <CardHeader title="My Contributions" />
        <CardContent>
          <Contributions />
        </CardContent>
        <CardActions>
          <Button disabled={!actionableSelection} onClick={handleEdit}>Edit</Button>
          <Button disabled={!actionableSelection}>Cancel</Button>
        </CardActions>
      </Card>
      <ContributionEdit />
    </>
  );
}

export default App;
