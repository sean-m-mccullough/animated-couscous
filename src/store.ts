import addDays from 'date-fns/addDays';
import subMonths from 'date-fns/subMonths';
import { configureStore } from '@reduxjs/toolkit';

import reducer from './reducers';
import preloadedState from './store.json';
import { Contribution, Contributions } from './types/contribution';
import { State } from './types/store';


const transformState = ({ contributions, dialogs }: any, startDate: Date): State => ({
  dialogs,
  selectedContribution: null,
  contributions: contributions.reduce((result: Contributions, { uuid, status, tfsa, rrsp }: Contribution, index: number) => ({
    ...result,
    [uuid]: {
      uuid,
      status,
      tfsa,
      rrsp,
      total: tfsa + rrsp,
      date: subMonths(startDate, index).toISOString(),
    }
  }), {})
});

export const createStore = (startDate: Date) => 
  configureStore({
    reducer,
    preloadedState: transformState(preloadedState, startDate) as any
  });

const store = createStore(addDays(new Date(), 15));

export default store;
