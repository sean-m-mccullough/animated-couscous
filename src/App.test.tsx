import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { afterEach, expect, it } from 'vitest'

import App from './App';
import { createStore } from './store';

const startDate = new Date(2020, 6, 3);

afterEach(cleanup)

it('renders with no issues', () => {
  const { container } = render(
    <Provider store={createStore(startDate)}>
      <App />
    </Provider>
  );

  expect(container).toMatchSnapshot();
});

it('opens a dialog when pending contribution is edited', async () => {
  render(
    <Provider store={createStore(startDate)}>
      <App />
    </Provider>
  );

  fireEvent.click(screen.getByText('July 3rd, 2020'));
  fireEvent.click(screen.getByText('Edit'));

  expect(document.body).toMatchSnapshot();
});
