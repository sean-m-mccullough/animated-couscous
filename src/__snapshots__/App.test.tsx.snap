// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`opens a dialog when pending contribution is edited 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
          >
            My Contributions
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <ul
          class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
        >
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                July 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                $
                1500
                <span
                  class="MuiBadge-root _root_25e320 css-1c32n2y-MuiBadge-root"
                >
                  <span
                    class="MuiBadge-badge _badge_25e320 MuiBadge-standard _standard_25e320 MuiBadge-anchorOriginTopRight _anchorOriginTopRight_25e320 MuiBadge-anchorOriginTopRightRectangular _anchorOriginTopRightRectangular_25e320 MuiBadge-overlapRectangular _overlapRectangular_25e320 MuiBadge-colorPrimary _colorPrimary_25e320 css-zza0ns-MuiBadge-badge"
                  >
                    PENDING
                  </span>
                </span>
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                June 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                May 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                April 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
          <div
            class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                March 3rd, 2020
              </p>
              <p
                class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
              >
                $
                1500
              </p>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </ul>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Edit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r1:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm css-1t1j96h-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r1:"
        >
          Edit Contribution
        </h2>
        <form
          action="#"
        >
          <div
            class="MuiDialogContent-root _root_0171e5 css-ypiqx9-MuiDialogContent-root"
          >
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r2:"
                id=":r2:-label"
              >
                RRSP
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                  >
                    $
                  </p>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r2:"
                  name="rrsp"
                  type="number"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      RRSP
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for=":r3:"
                id=":r3:-label"
              >
                TFSA
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1q6at85-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                  >
                    $
                  </p>
                </div>
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r3:"
                  name="tfsa"
                  type="number"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      TFSA
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Cancel
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Accept
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`renders with no issues 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
        >
          My Contributions
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              July 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              $
              1500
              <span
                class="MuiBadge-root _root_25e320 css-1c32n2y-MuiBadge-root"
              >
                <span
                  class="MuiBadge-badge _badge_25e320 MuiBadge-standard _standard_25e320 MuiBadge-anchorOriginTopRight _anchorOriginTopRight_25e320 MuiBadge-anchorOriginTopRightRectangular _anchorOriginTopRightRectangular_25e320 MuiBadge-overlapRectangular _overlapRectangular_25e320 MuiBadge-colorPrimary _colorPrimary_25e320 css-zza0ns-MuiBadge-badge"
                >
                  PENDING
                </span>
              </span>
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              June 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              May 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              April 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
        <div
          class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              March 3rd, 2020
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
            >
              $
              1500
            </p>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
    <div
      class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Edit
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
